10_rest/willshen8/pkg/puppy/syncstore.go

```diff
@@ -0,0 +1,75 @@
+package puppy
+
+import (
+	"strconv"
+	"sync"
+)
+
+type SyncStore struct {
+	syncStore sync.Map
+	nextID    uint32
+	mu        sync.Mutex
+}
+
+// NewMapStore initialise a new SyncStore
+func NewSyncStore() *SyncStore {
+	return &SyncStore{syncStore: sync.Map{}}
+}
+
+// CreatePuppy create a new puppy and store in mapStore.
+func (s *SyncStore) CreatePuppy(p *Puppy) (uint32, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if i, err := strconv.Atoi(p.Value); err == nil {
+		if i < 0 {
+			return 0, Errorf(ErrorInvalidInput, "Puppy value can't be negative.")
+		}
+	} else {
+		return 0, Errorf(ErrorInvalidInput, "Unrecognised input.")
+	}
+	s.nextID++
+	p.ID = s.nextID
+	s.syncStore.Store(p.ID, *p)
+
+	return p.ID, nil
+}
+
+// ReadPuppy read a puppy given its id.
+// It returns the pointer to that puppy.
+func (s *SyncStore) ReadPuppy(id uint32) (*Puppy, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if p, ok := s.syncStore.Load(id); ok {
+		returnPuppy := p.(Puppy)
+		return &returnPuppy, nil
+	}
+	return nil, Errorf(ErrorNotFound, "Puppy ID can't be found.")
+}
+
+// UpdatePuppy updates the store with key of id with the new puppy.
+func (s *SyncStore) UpdatePuppy(id uint32, p *Puppy) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
you are swallowing error here.
what if `value` is not a number?

```bash
curl localhost:8888/api/puppy/1 -X PUT -d '{"value": "aaaaaaaaaaaaaaa"}'
"Puppy is updated successfully."
curl localhost:8888/api/puppy/1
{"id":1,"breed":"","color":"","value":"aaaaaaaaaaaaaaa"}

```

Reply:
Not going to change sorry, initially the thought is to be able to update as multiple currency values as a string. But the lab has requirement of value less than 0. So have to make compromises.

Reply:
Not going to change sorry, initially the thought is to be able to update as multiple currency values as a string. But the lab has requirement of value less than 0. So have to make compromises.

Reply:
@n0npax Thanks for your help so far. I've added the change to make it consistent. My initial thought was to make the field bit flexible so that people can enter things like $10, £10, etc. I guess it will be an extension of the project.